[[ts-elasticsearch]]
==== {es} issues and solutions

[discrete]
[[ts-elasticsearch-409]]
===== {es} rejects documents with `version_conflict_engine_exception`

*Symptoms*

When elastic-agent transmits events to {ls} and {ls} to {es} with <<plugins-outputs-elasticsearch,{es} output plugin>>, {es} may reject documents with `version_conflict_engine_exception` exception.
The logs may look like as follows:

-----
[2024-06-04T12:57:59,670][WARN ][logstash.outputs.elasticsearch][elastic-agent-to-logstash][abcd1....] Failed action {:status=>409, :action=>["create",
{:_id=>nil, ... {"error"=>{"type"=>"version_conflict_engine_exception", "reason"=>"[Cx8OD9UBQrQcJL5Jv3oJSE5iHUs=]: version conflict, document already exists (current version [1])",
"index_uuid"=>"-h1tlFaSQK-2plBj_kNTig", "shard"=>"0", "index"=>".ds-logs-m365_defender.log-2024.06.02-000021"}}}}
-----

*Background*

For the most cases, underlining data stream has an integration includes fingerprint processor which calculates document `_id`. This can be confirmed by visiting ingest pipelines list in {kb} and
filtering out with data stream pattern. For the particular error case shown above, it will be `logs-m365_defender.log-{version}` ingest pipeline.

*Possible case solutions*

- Some integrations convert the event into JSON shape and JSON fields are used in fingerprinting. Make sure that all JSON fields are available during the `_id` has calculation to provide a persistent granularity.
- If Logstash is facing a backpressure, it cannot ack the events back to elastic-agent. Agent can throw timeout and resend the event. However, in a reality event is already indexed into {es}. Extending timeout in elastic-agent may help.